src/GraphQL/Subscription/ObservableExtensions.cs

```diff
@@ -0,0 +1,253 @@
+                        await ProcessErrorAsync(queueData.Error!).ConfigureAwait(false);
+                    }
+                    else if (queueData.Type == QueueType.Completion)
+                    {
+                        if (!_token.IsCancellationRequested)
+                            _observer.OnCompleted();
+                    }
+                    // once the event has been passed along, dequeue it
+                    lock (_queue)
+                    {
+                        _ = _queue.Dequeue();
+                        queueData = _queue.Count > 0 ? _queue.Peek() : null;
+                    }
+                    // if the queue is empty, immedately quit the loop, as any new
+                    // events queued will start ReturnDataAsync
+                }
+            }
+
+            /// <summary>
+            /// Wait for the transform to complete and push the data (or error) back to the observer.
+            /// If the observer has been disposed, then data and errors are ignored.
+            /// </summary>
+            private async Task ProcessDataAsync(Task<TOut> dataTask)
+            {
+                if (_token.IsCancellationRequested)
+                    return;
+                TOut dataOut;
+                try
+                {
+                    dataOut = await dataTask.ConfigureAwait(false);
+                }
+                catch (Exception error)
+                {
+                    if (!_token.IsCancellationRequested)
+                        _observer.OnError(error);
+                    return;
+                }
+                if (!_token.IsCancellationRequested)
+                    _observer.OnNext(dataOut);
+            }
+
+            /// <summary>
+            /// Wait for the transform to complete and push the error back to the observer.
+            /// If the observer has been disposed, then errors are ignored.
+            /// </summary>
+            private async Task ProcessErrorAsync(Task<Exception> errorTask)
+            {
+                if (_token.IsCancellationRequested)
+                    return;
+                Exception errorOut;
+                try
+                {
+                    errorOut = await errorTask.ConfigureAwait(false);
+                }
+                catch (Exception error)
+                {
+                    if (!_token.IsCancellationRequested)
+                        _observer.OnError(error);
+                    return;
+                }
+                if (!_token.IsCancellationRequested)
+                    _observer.OnError(errorOut);
+            }
+
+            /// <summary>
+            /// Disposes of the underlying observable sequence
+            /// </summary>
+            public void Dispose()
+            {
+                // cancel pending operations and prevent pending operations
+                // from returning data after the observable has been detatched
+                _cancellationTokenSource?.Cancel();
```

Comment:
I would like to switch dispose logic here to one used by Rx and `SampleObservable` to eliminate any possible concurrency case of double disposition:
```
var cts = Interlocked.Exchange(ref _cancellationTokenSource, null);
if (cts == null)
    return;
// Other code from Dispose
```

Reply:
Done